Healthcheck middleware tests endpoints for availability

JoeChapman opened the pull request:
Mountable health check middleware makes requests after configurable intervals for the availability of a web service.

JoeChapman committed:
Healthcheck middleware tests endpoints for availability

- mountable healthcheck middleware makes requests after configurable intervals.

lennym commented on lib/index.js:
```diff
@@ -0,0 +1,48 @@
+      // Process all pending requests
+      return Promise.all(requests)
+        .then(resolved => {
+          req.log(resolved);
```
I believe `req.log` expects a log level as a first argument.

lennym commented on lib/index.js:
```diff
@@ -0,0 +1,48 @@
+
+      // Process all pending requests
+      return Promise.all(requests)
+        .then(resolved => {
```
Does this just assume that if the healthcheck endpoints return at all then they are healthy? How would it handle a status endpoint which always responded, but responded with 2xx/5xx status codes?

JoeChapman replied:
yes, this is true. As an initial implementation this is assumed.

lennym replied:
I don't think that's a valid assumption. I would expect any (competent) healthcheck endpoint to be able to successfully report its own lack of health.

For example, https://status.github.com/ will respond even when the rest of github is on fire.

easternbloc replied:
I don't think we can rely on just a response here. As lenny states we should rely on the status code when we get a response. It's the norm.

lennym commented:
I think there's a slight issue with this that the state of an endpoint is not cached, and so if one request fetches endpoints and they are unhealthy, the next request will find all the previous requests cached, not do anything, and pass though as if all is ok.

JoeChapman commented:
> I think there's a slight issue with this that the state of an endpoint is not cached, and so if one request fetches endpoints and they are unhealthy, the next request will find all the previous requests cached, not do anything, and pass though as if all is ok.

@lennym do you mean that the state of an endpoint _is_  ~not~ cached

easternbloc reviewed (changes requested):
We need to read the status code on a successful response.

JoeChapman commented on lib/index.js:
```diff
@@ -0,0 +1,65 @@
+            try {
+              response = JSON.parse(resolved[i]);
+            } catch (e) {
+              req.log('debug', response, e);
```
Perhaps it'll make more sense here to check for non-JSON responses i.e.
```
response = response[i];
```

JoeChapman commented:
@easternbloc @lennym I've updated the response handler. Please use your 👀 to l=review the code.

easternbloc commented on lib/index.js:
```diff
@@ -0,0 +1,65 @@
+
+            if (response) {
+              const status = response.statusCode || response.status;
+              if (status.startsWith(5)) {
```
why not just check for a successful 200 there are more errors than 500s that you could get back. This is flakey.

JoeChapman commented:
@easternbloc please take another gander

lennym commented on lib/index.js:
```diff
@@ -0,0 +1,72 @@
+              // If there is a status code
+              // but it's not a 200, throw an Error
+              if (status) {
+                if (parseInt(status, 10) !== 200) {
```
I'd check for a 2xx rather than a strict 200. It would be reasonable for a healthcheck to respond with 204 for example. Arguably a 304 *might* be legit too.

FWIW, this is how hof-model detects a "success" - https://github.com/UKHomeOfficeForms/hof-model/blob/master/lib/model.js#L179

easternbloc replied:
I don't think 300's should be in. If we're getting a 300 we should be following the redirect no? In which case you'd expect the end result to be a 2xx.

I do agree on 200's being in though.

lennym replied:
Does `request` follow redirects automatically?

easternbloc replied:
https://github.com/request/request/blame/master/README.md#L16

lennym commented:
@JoeChapman Re caching: what I mean is that only the request times are cached, and are done so before the status is resolved.

This means that if a service is down, and the healthcheck fails, if the user then refreshes immediately (as they are likely to) then this middleware will look at the cache, see that the health was just checked a few seconds ago, and pass through successfully, without regard for the fact that the healthcheck is failing.

We need to cache both the time and the state of the last request, so that if there was a recent *failed* check then the cached state is still a failure.

lennym commented:
Actually the best thing to do is probably to cache the *promise*, since that way even in-flight requests can still be re-used for subsequent checks.

That is [here](https://github.com/UKHomeOfficeForms/hof-middleware-healthcheck/pull/1/files#diff-6d186b954a58d5bb740f73d84fe39073R23) we push a new promise if none exists in the cache, and the cached promise if it does.

JoeChapman committed:
throw error on >= 400

JoeChapman committed:
Cache promises and reset all promises on resolve and reject

So in-flight requests are not lost if a new request interrupts
the current request, promises are cached so they can be used again

JoeChapman committed:
add travis badge
